libgalois/include/katana/PropertyGraph.h

```diff
@@ -842,6 +903,633 @@ CreateSymmetricGraph(PropertyGraph* pg);
+  EdgeIndexToLabelMap edge_index_to_label_map_;
+  uint32_t num_unique_edge_labels_ = 0u;
+};
+
+// TODO(amber): make OrigEdgeIDMap optional via template argument
+class KATANA_EXPORT EdgeShuffleTopology : public GraphTopology {
+  using Base = GraphTopology;
+  using OrigEdgeIDMap = katana::NUMAArray<Edge>;
+  using EdgeLabelTy = PropertyGraph::TypeSetID;
+
+public:
+  EdgeShuffleTopology() = default;
+  EdgeShuffleTopology(EdgeShuffleTopology&&) = default;
+  EdgeShuffleTopology& operator=(EdgeShuffleTopology&&) = default;
+
+  EdgeShuffleTopology(const EdgeShuffleTopology&) = delete;
+  EdgeShuffleTopology& operator=(const EdgeShuffleTopology&) = delete;
+
+  static EdgeShuffleTopology MakeTransposeCopy(const PropertyGraph* pg);
+  static EdgeShuffleTopology MakeOriginalCopy(const PropertyGraph* pg);
+
+  auto original_edge_id(Edge eid) const noexcept {
+    KATANA_LOG_DEBUG_ASSERT(eid < orig_edge_ids_.size());
+    return orig_edge_ids_[eid];
+  }
+
+  void SortEdgesByLabelThenDst() noexcept;
+
+protected:
+  NUMAArray<Edge>&& MoveOrigEdgeIDs() noexcept {
+    return std::move(orig_edge_ids_);
+  }
+
+  NUMAArray<Edge>& GetOrigEdgeIDs() noexcept { return orig_edge_ids_; }
+
+private:
+  // Needed for creating EdgeLabelAwareTopology from EdgeShuffleTopology
+  friend class EdgeLabelAwareTopology;
+
+  EdgeShuffleTopology(
+      const PropertyGraph* pg, NUMAArray<Edge>&& adj_indices,
+      NUMAArray<Node>&& dests, NUMAArray<Edge>&& orig_edge_ids) noexcept
+      : GraphTopology(std::move(adj_indices), std::move(dests)),
+        prop_graph_(pg),
+        orig_edge_ids_(std::move(orig_edge_ids)) {}
+
+  const PropertyGraph* prop_graph_;
+  OrigEdgeIDMap orig_edge_ids_;
+};
+
+/// store adjacency indices per each node such that they are divided by edge label type.
+/// Requires sorting the graph by edge label type
+class KATANA_EXPORT EdgeLabelAwareTopology : public GraphTopologyTypes {
+  // TODO(amber): add a template flag to choose between copying edge labels from
+  // PropertyGraph vs accessing the data in PropertyGraph through
+  // orig_edge_ids_
+
+protected:
+  using EdgeLabelTy = PropertyGraph::TypeSetID;
+
+public:
+  EdgeLabelAwareTopology() = default;
+  EdgeLabelAwareTopology(EdgeLabelAwareTopology&&) = default;
+  EdgeLabelAwareTopology& operator=(EdgeLabelAwareTopology&&) = default;
+
+  EdgeLabelAwareTopology(const EdgeLabelAwareTopology&) = delete;
+  EdgeLabelAwareTopology& operator=(const EdgeLabelAwareTopology&) = delete;
+
+  static EdgeLabelAwareTopology MakeFromDefaultTopology(
+      const PropertyGraph* pg, const EdgeLabelIndex* edge_label_idx);
+  static EdgeLabelAwareTopology MakeFromTransposeTopology(
+      const PropertyGraph* pg, const EdgeLabelIndex* edge_label_idx);
+
+  uint64_t num_nodes() const noexcept {
+    return adj_indices_.size() / edge_label_index_->num_unique_labels();
+  }
+
+  uint64_t num_edges() const noexcept { return dests_.size(); }
+
+  // Edge accessors
+
+  /**
```

Comment:
can we stick with one style of Doxygen comments in a file? `///`

Reply:
done